### PR TITLE
feat: oriented poles for off-diagonal (rotated/monoclinic) dispersion

### PR DESCRIPTION
**Pull request by MarkCunningham0410**
This adds an optional orientation to dispersion poles, enabling media whose permittivity tensor has frequency-dependent **off-diagonal** components — rotated/tilted crystals and monoclinic media (shear phonon polaritons, ghost waves), where the optical axes do not align with the grid. It builds directly on the per-axis (diagonal) dispersion from #402, which required optical axes to be grid-aligned; that restriction is now lifted for the dispersive part of the tensor.

A pole with an `orientation` unit vector becomes a 1D oscillator along that direction, with coupling tensor `K·uuᵀ`:

```python
# oscillator along the (1, 1, 0) diagonal — e.g. one axis of a rotated crystal
material = fdtdx.Material(
    permittivity=2.0,
    dispersion=fdtdx.DispersionModel(poles=(
        fdtdx.LorentzPole(
            resonance_frequency=w_to,
            damping=g,
            delta_epsilon=de,
            orientation=(1.0, 1.0, 0.0),  # normalized internally
        ),
    )),
)
```

For the common tilted-crystal case, `DispersionModel.rotated(R)` converts a per-axis (diagonal) model into an oriented one by rotating each pole's principal axes; signed axis permutations are recognized and stay diagonal (no cost).

Complex full-tensor constructors round out the API: `Material.from_complex_permittivity` accepts flat 9-tuples and nested 3×3 complex tensors (real part → permittivity tensor, imaginary part → conductivity tensor component-wise via σ_ij = ωε₀ε″_ij, exact at the reference frequency; invertibility of the real part is validated). `from_loss_tangent` supports flat 9-tuples; `from_refractive_index` rejects tensors (matrix vs elementwise square is ambiguous).

## Implementation notes

- The dispersive coupling arrays `c3`/`c4` gain a 9-component tier (3×3 per pole), extending the 1/3 sizing from #402 and mirroring the existing 1/3/9 material array sizing.
- The fully anisotropic update kernel gains the ADE polarization step; the off-diagonal delta is folded into the curl before the existing Yee averaging, so there is no extra linear solve per step.
- Off-diagonal coupling terms at material boundaries are **symmetrized**: the pair weight between a target location and a neighboring sample is `stencil · (c(r)m(s) + m(r)c(s))/2` with `m = (c ≠ 0)` the material indicator, keeping the coupling blocks of the update operator exact mutual adjoints where the coefficients vary and restricting coupling to material cells. With the plain `c·avg(E)` form, a boundary cell couples to vacuum neighbors that do not couple back, and the resulting non-normal operator amplifies boundary modes in media with off-diagonal coupling and mixed-sign permittivity. In homogeneous regions it reduces to the plain form (bulk behavior unchanged). Non-uniform grids keep the legacy averaging for now (the weighted symmetrization needs matching edge weights).
- Oriented dispersion forces the 9-component permittivity tier and supports only the **checkpointed** gradient method — the reversible path raises a descriptive error at initialization and in `run_fdtd`.
- Setup-time sampling helpers (effective permittivity, spectra) are generalized to mixed component tiers.

## Two fixes found while testing

- The broadband TFSF impedance filter allocated its full spectrum (`num_time_steps × components × slice cells`) even when the source plane sits in a non-dispersive material — all-zero coupling coefficients in the slice mean ε(ω) is flat and the filter is the identity, so it is now skipped. This is the complementary case to #412 (which is about the filter being component-averaged when the slice *is* anisotropically dispersive) and does not change that behavior — the setup warning from #402 still fires there.
- `GaussianPlaneSource` used an xy-indexed meshgrid that swaps the horizontal/vertical coordinates on non-square source planes, misplacing the Gaussian spot; once the misplaced center was more than a radius off the plane, the profile was all zeros and energy normalization produced NaN fields. Square planes (all existing tests/examples) were unaffected. Now ij-indexed in (width, height) order.

## Validation

- An x-oriented pole reproduces the per-axis (diagonal) reference — the arithmetic is identical.
- A 45°-rotated dispersive crystal matches the grid-aligned crystal probed at 45° polarization.
- A monoclinic non-orthogonal oscillator pair (two poles with non-orthogonal orientations) runs stably.
- A rotated lossy crystal exercises the full-tensor σ_E update path end to end via `from_complex_permittivity`.
- Checkpointed gradients are finite and reach the off-diagonal coupling entries.
- Full unit + integration suite passes; isotropic and per-axis setups produce identical arrays.

Example: `examples/hyperbolic_asymmetric_refraction.py` — calcite driven inside its Reststrahlen band with the optic axis tilted 25° above the surface via `DispersionModel.rotated`. Two mirrored ±45° beams refract asymmetrically (+23° positive vs +53° negative) because the frequency-dependent off-diagonal ε_xz shears the iso-frequency curve — a signature that is impossible both for a static tensor (mixed-sign eigenvalues are unstable) and for the diagonal dispersion of #402 (mirror-symmetric refraction). The script plots the lab-frame tensor components over the band and checks the measured energy-corridor angles against the analytic group-velocity directions (0.1° and 4.0° agreement).

Out of scope: symmetrized off-diagonal averaging on non-uniform grids (legacy form kept there); reversible autodiff through oriented dispersion.

Related: #402, #75, #204, #4, #412.

## Summary by CodeRabbit

* **New Features**
  * Added oriented (off-diagonal) dispersive materials with full tensor permittivity support, including rotation/orientation handling.
  * Exposed `compute_pole_coefficients_tensor` in the public API and added a GPU hyperbolic asymmetric refraction example.
* **Bug Fixes**
  * Fixed Gaussian plane-source spot placement on non-square planes.
  * Improved dispersive broadband filtering by avoiding extra processing when coupling is absent.
* **Documentation**
  * Expanded oriented-dispersion guidance, including checkpointed-only gradient restrictions.
* **Tests**
  * Added extensive coverage for oriented/off-diagonal behavior, complex tensor constructors, and gradient-safety/runtime checks.

### CODE REVIEW

**Comment by MarkCunningham0410** [comment]
Converted this to draft so I can get to the review comments some stage soon.

File: .claude/skills/fdtdx/SKILL.md [hunk 1]
```diff
@@ -152,14 +152,16 @@ Stability needs `γ·dt < 2`; physically `γ·dt ≪ 1`, so `c2 ≈ −1` and th
...
+
 **ArrayContainer fields** (all `None` unless any object is dispersive):
 - `dispersive_P_curr`, `dispersive_P_prev` — shape `(num_poles, 3, Nx, Ny, Nz)`, field-dtype (complex if `use_complex_fields`). Not differentiable (state-only; `None` cotangent in both gradient paths).
 - `dispersive_c1`, `dispersive_c2`, `dispersive_c3` — shape `(num_poles, C, Nx, Ny, Nz)` with `C = 1` when all dispersion is isotropic (middle axis broadcasts over field components) or `C = 3` for per-axis dispersion (gated by `ObjectContainer.all_objects_isotropic_dispersion`). Config dtype. Differentiable: cotangents flow through them in both the reversible and checkpointed paths.
```
[the remark this answers is not in the document]
Comment (MarkCunningham0410) [reply .claude/skills/fdtdx/SKILL.md]: The bullet now documents the 9-component c3/c4 coupling tier and its gate `(all_objects_axis_aligned_dispersion)`, alongside the 1/3 tiers for c1/c2.

File: src/fdtdx/dispersion.py [hunk 2]
```diff
@@ -745,8 +1110,12 @@ def compute_eps_spectrum_from_coefficients(
...
     # Average the material-component axis (identity for C = 1), mirroring the
     # eps_inf reduction above — this scalar spectrum feeds an impedance filter
-    # that has no notion of polarization.
-    chi_per_cell = chi_per_cell.mean(axis=1)  # → (M, *spatial)
+    # that has no notion of polarization. For a 9-component coupling only the
+    # diagonal entries carry impedance information.
+    if chi_per_cell.shape[1] == 9:
+        chi_per_cell = chi_per_cell[:, (0, 4, 8)].mean(axis=1)
+    else:
+        chi_per_cell = chi_per_cell.mean(axis=1)  # → (M, *spatial)
```
[the remark this answers is not in the document]
Comment (MarkCunningham0410) [reply src/fdtdx/dispersion.py:1172]: Added a finiteness check on the components and on the norm, so NaN, inf, and overflow-to-inf (e.g. 1e200 components) all raise at construction. Unit test covers the three cases.

File: src/fdtdx/dispersion.py [hunk 3]
```diff
@@ -108,8 +207,41 @@ class Pole(TreeClass, ABC):
...
+        directly on the class, not inherited)."""
+        if self.orientation is None:
+            return
+        vec = self.orientation
+        if not isinstance(vec, tuple) or len(vec) != 3:
+            raise ValueError(f"Pole orientation must be a 3-tuple (x, y, z), got {vec!r}.")
+        norm = float(np.sqrt(float(vec[0]) ** 2 + float(vec[1]) ** 2 + float(vec[2]) ** 2))
+        if norm < 1e-12:
+            raise ValueError("Pole orientation must be a non-zero vector.")
+        object.__setattr__(self, "orientation", (float(vec[0]) / norm, float(vec[1]) / norm, float(vec[2]) / norm))
```
[the remark this answers is not in the document]
Comment (MarkCunningham0410) [reply src/fdtdx/dispersion.py]: Added a finiteness check on the components and on the norm, so NaN, inf, and overflow-to-inf (e.g. 1e200 components) all raise at construction. Unit test covers the three cases.

File: src/fdtdx/dispersion.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (MarkCunningham0410) [reply src/fdtdx/dispersion.py:1172]: Sorry, typed it into the wrong one. The 9-component path now reconstructs each cell's 3×3, inverts it, and averages the diagonal of the actual `ε∞` tensor instead of reciprocating the inverse's diagonal. Added a unit test against `susceptibility_tensor` with an off-diagonal `ε∞`.

File: src/fdtdx/fdtd/update.py [hunk 4]
```diff
@@ -286,6 +285,84 @@ def update_E(
...
+                            cij_pad = jnp.pad(cij, pad, mode="edge")
+                            mask_pad = (cij_pad != 0.0).astype(cij.dtype)
+                            row = row + 0.5 * (
+                                cij * _avg_offdiag(mask_pad * E_pad[j], component=j, location=i)
+                                + (cij != 0.0) * _avg_offdiag(cij_pad * E_pad[j], component=j, location=i)
+                            )
+                        else:
+                            # non-uniform grids keep the legacy form for now (the
+                            # weighted symmetrization needs matching edge weights)
+                            row = row + cij * avg_at[(i, j)]
```
[the remark this answers is not in the document]
Comment (MarkCunningham0410) [reply src/fdtdx/fdtd/update.py]: Went with the reject option: oriented poles on a non-uniform grid now raise `NotImplementedError` at initialization, next to the existing reversible/CCPR guards, and the legacy branch is removed from the kernel so the symmetrized stencil is unconditional. Per-axis (diagonal) dispersion on non-uniform grids is unaffected. Integration test added.

File: src/fdtdx/materials.py [hunk 5]
```diff
@@ -451,6 +472,13 @@ def from_complex_permittivity(
         omega = _resolve_reference_omega(reference, wavelength, frequency)
         eps_real, sigma_e = _split_complex_property(permittivity, omega, constants.eps0)
         mu_real, sigma_m = _split_complex_property(permeability, omega, constants.mu0)
+        eps_mat = np.array(_normalize_material_property(eps_real), dtype=np.float64).reshape(3, 3)
+        det = float(np.linalg.det(eps_mat))
+        if abs(det) < 1e-9 * max(1.0, float(np.abs(eps_mat).max()) ** 3):
+            raise ValueError(
+                "The real part of the complex permittivity tensor is singular (determinant ~ 0). "
+                "FDTDX stores the inverse permittivity, so the real part must be an invertible 3x3 tensor."
+            )
```
[the remark this answers is not in the document]
Comment (MarkCunningham0410) [reply src/fdtdx/materials.py]: The determinant check now runs over both the ε and μ real parts; a singular μ raises with the matching message. Test added.

File: src/fdtdx/materials.py [hunk 6]
```diff
@@ -861,39 +912,59 @@ def compute_allowed_dispersive_coefficients(
...
+    diag_entries = (0, 4, 8)
     for m_idx, (_, mat) in enumerate(ordered):
         if mat.dispersion is None:
             continue
         if num_components == 1 and not mat.has_isotropic_dispersion:
             raise ValueError("num_components=1 requires isotropic dispersion, but a material has per-axis poles.")
+        if coupling_components < 9 and not mat.has_axis_aligned_dispersion:
+            raise ValueError(
+                "coupling_components < 9 requires axis-aligned dispersion, but a material has oriented poles."
+            )
```
[the remark this answers is not in the document]
Comment (MarkCunningham0410) [reply src/fdtdx/materials.py:955]: Added the explicit raise. Initialization can't currently produce this combination (`coupling_components` is 1 only when all dispersion is isotropic), but the function-local invariant is now enforced rather than assumed.

File: src/fdtdx/objects/sources/tfsf.py [hunk 7]
```diff
@@ -91,6 +91,12 @@ def _build_dispersive_H_filter(
     c4_np = None if c4_slice is None else np.asarray(c4_slice)
     inv_eps_inf_np = np.asarray(inv_eps_inf_slice)
 
+    # All-zero coupling in the source slice (non-dispersive material at the
+    # source plane) means eps(omega) is flat and the filter is the identity —
+    # skip it rather than allocate the full broadband spectrum.
+    if c3_np.size == 0 or (not np.any(c3_np) and (c4_np is None or not np.any(c4_np))):
+        return jnp.asarray(raw_samples, dtype=dtype)
```
[the remark this answers is not in the document]
Comment (MarkCunningham0410) [reply src/fdtdx/objects/sources/tfsf.py:92]: Reordered - the identity check now runs before the FFT grid is built. For scale: the pre-return allocation was the `rfft` frequency grid (~256 KB at 16k steps), not the broadband spectrum itself, which was already skipped - but the reorder costs nothing.

File: src/fdtdx/dispersion.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (MarkCunningham0410) [reply src/fdtdx/dispersion.py]: Switched to scale-first normalization (divide by the largest |component| before the norm), so (1e200, 1e200, 0) now normalizes to (√½, √½, 0); added that case plus a denormal-scale one to the unit test. Non-finite components still raise.

[the remark this answers is not in the document]
Comment (MarkCunningham0410) [reply src/fdtdx/dispersion.py]: check again

**Comment by ymahlau** [comment]
@MarkCunningham0410 Thanks for the amazing PR and sorry for ignoring it a while, I had a short vacation. Please let me know when you think the PR is at a point where I should take a more detailed look at the code.

The general structure from skimming the changes looks good already. Very excited to see this feature realized!

**Comment by MarkCunningham0410** [comment]
Thank you! It's good to go now, I have potentially another PR coming up soon for gyrotropic - but it's a bit more difficult for me to confirm the validity of that since my PhD didn't really go that far. But the contents of this PR are ready!

File: src/fdtdx/fdtd/update.py [hunk 8]
```diff
@@ -286,6 +285,73 @@ def update_E(
...
+                # the uniform 4-point stencil is always valid here.
+                rows = []
+                for i in range(3):
+                    row = disp_c3[:, 3 * i + i] * arrays.fields.E[i]
+                    for j in range(3):
+                        if j == i:
+                            continue
+                        cij = disp_c3[:, 3 * i + j]
+                        pad = ((0, 0), (1, 1), (1, 1), (1, 1))
+                        cij_pad = jnp.pad(cij, pad, mode="edge")
```
Comment (ymahlau) [inline src/fdtdx/fdtd/update.py]: The padding here might disagree with the paddings on the fields like E_pad (which use pml/periodic boundary aware "pad_fields_for_boundary"). Could that lead to inconsistencies and numerical errors?

**Comment by ymahlau** [comment]
@MarkCunningham0410 I took a closer look at the code and have very little comments, it looks very good already. Only a minor padding issue caught my eye, but I am not sure if that is even a problem in practice (see above)

File: src/fdtdx/fdtd/update.py (hunk 8 above)
Comment (MarkCunningham0410) [reply src/fdtdx/fdtd/update.py]: I think a PML is unaffected. The field is zero there, so the coefficient padding value gets multiplied by zero anyway. You could be right about a periodic boundary though, as an off-diagonal material could use the wrong neighbour coefficient as the periodic boundary wraps around it. This could probably be a quick claude fix, maybe wrapping the coefficient padding on periodic axes?

**Comment by MarkCunningham0410** [comment]
@ymahlau the padding fixed from above is pushed now, the coefficient padding wraps on periodic axes and edge-replicates elsewhere. Added a unit test and everything else is bit-identical.

**Comment by ymahlau** [comment]
Thanks a lot for this amazing PR!
